leshan-core/src/main/java/org/eclipse/leshan/core/attr/AttributeSet.java

```diff
@@ -0,0 +1,77 @@
+package org.eclipse.leshan.core.attr;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.leshan.core.request.exception.InvalidRequestException;
+
+/**
+ * A collection of {@link Attribute} instances that are handled as a unit regarding validation etc.
+ */
+public class AttributeSet {
+    
+    private static final String PARAM_SEP = "&";
+    private static final String CANCEL = "cancel";
+
+    private Map<AttributeName, Attribute> attributeMap = new HashMap<>();
+    private boolean cancel;
+
+    public AttributeSet(boolean cancel) {
+        this.cancel = cancel;
+    }
+    
+    public AttributeSet(Attribute...attributes) {
+        this(false);
+        for (Attribute attr : attributes) {
+            // Check for duplicates
+            if (attributeMap.containsKey(attr.getName())) {
+                throw new InvalidRequestException("Cannot create attribute set with duplicates (attr: '%s')",
+                        attr.getCoRELinkParam()); 
+            }
+            attributeMap.put(attr.getName(), attr);
+        }
+    }
+
+    public void validate(AssignationLevel assignationLevel) {
+        // Can all attributes be assigned to this level?
+        for (Attribute attr : attributeMap.values()) {
+            if (!attr.canBeAssignedTo(assignationLevel)) {
+                throw new InvalidRequestException("Attribute '%s' cannot be assigned to level %s",
+                        attr.getCoRELinkParam(), assignationLevel.name());
+            }
+        }
+        Attribute pmin = attributeMap.get(AttributeName.MINIMUM_PERIOD);
+        Attribute pmax = attributeMap.get(AttributeName.MAXIMUM_PERIOD);
+        if ((pmin != null) && (pmax != null) && (Integer) pmin.getValue() > (Integer) pmax.getValue()) {
+            throw new InvalidRequestException("Cannot write attributes where '%s' > '%s'",
+                    pmin.getCoRELinkParam(), pmax.getCoRELinkParam());
+        }
+    }
+    
+    public Attribute[] getAttributes() {
+        return attributeMap.values().toArray(new Attribute[attributeMap.size()]);
+    }
+    
+    public String[] toQueryParams() {
```

Comment:
Maybe we don't really need to make it public.
We can maybe merge it with toString, but we will see later depending on how we are using AttributeSet.

Reply:
It's used for ObserveSpec in https://github.com/eclipse/leshan/blob/master/leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CoapRequestBuilder.java so I'll leave it public for now.